### PR TITLE
Use consistent package naming for HEAD.nix files

### PR DESCRIPTION
**Pull request by rht**
###### Motivation for this change

This is a continuation of https://github.com/NixOS/nixpkgs/pull/21822#issuecomment-274345923. Naming is based on https://nixos.org/nixpkgs/manual/#sec-package-naming.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by rht** [comment]
> The Haskell compiler has a very precise version number and I want the Nix expression to use that particular version, too.

So does the Coq compiler, and I had requested for the precise version taken from the git tags, but the disambiguation convention is to use commit date, which is better for scaling and automation.

**Comment by Profpatsch** [comment]
Yes, it’s not a good idea to make version tags consistent, especially not for very inconsistent and project-dependent topics like head versions.

**Comment by rht** [comment]
There are various head versions convention: projectname-numberofcommits-commithash (autogenerated in any git repo), numberofcommits (svn), ... (I forgot how hg does it, as it might include date).

There are two ways to be consistent in head versions:
1. projectname-unstable-YYYY-MM-DD
2. follow upstream's convention. If upstream uses git/darcs/bzr, then version is named using git/darcs/bzr convention, ... 

IMO, it should be either of the two options, but not a mixture of them.

**Comment by jagajaga** [comment]
@rht the reason to this is because nix won't handle non numbers as a version. So `foo-ars213h` and `foo-1233.33` are different pkgs for it.

**Comment by jagajaga** [comment]
@peti my fault.

File: pkgs/development/compilers/ghc/head.nix [hunk 1]
```diff
@@ -12,7 +12,7 @@ let
 
   commonBuildInputs = [ ghc perl autoconf automake happy alex python3 ];
 
-  version = "8.1.20170106";
+  version = "2017-01-06";
```
Comment (rht) [inline pkgs/development/compilers/ghc/head.nix]: Addressing @peti's objection: the git-generated version is `ghc-8.3-start-421-g43a31683ac`; the revision `b4f2afe70ddbd0576b4eba3f82ba1ddc52e9b3bd` doesn't have a tag, so I assume this revision was the HEAD commit of the time. And I see that "8.1.20170106" is a convention specific to ghc HEAD/unstable package rather than nix in general.

Comment (rht) [inline pkgs/development/compilers/ghc/head.nix]: For reference, here are the recent tags in ghc:
```
ghc-8.0.1-rc1
ghc-8.0.1-rc2
ghc-8.0.1-rc3
ghc-8.0.1-rc4
ghc-8.0.1-release
ghc-8.0.2-rc1
ghc-8.0.2-rc2
ghc-8.0.2-release
ghc-8.1-start
ghc-8.2.1-rc1
ghc-8.2.1-rc2
ghc-8.3-start
```

Comment (peti) [inline pkgs/development/compilers/ghc/head.nix]: @rht, I don't understand what you're getting at. Could you please revert these changes? If you have a better suggestion to be used for GHC HEAD version numbers, then by all means please make a suggestion in the form of a new PR, but the version numbers we have now are unsuitable, IMHO.

Comment (rht) [inline pkgs/development/compilers/ghc/head.nix]: The ghc HEAD version numbers as used by the upstream (generated by git, e.g. ghc-8.3-start-421-g43a31683ac) are not used at all here.

To explain once again, my change is based on
> If a package is not a release but a commit from a repository, then the version part of the name must be the date of that (fetched) commit. The date must be in "YYYY-MM-DD" format. Also append "unstable" to the name - e.g., "pkgname-unstable-2014-09-23"

from https://nixos.org/nixpkgs/manual/#sec-package-naming.

Comment (peti) [inline pkgs/development/compilers/ghc/head.nix]: Please feel free to suggest better version numbers for the ghc and ghcjs expressions in a new PR. Then we can discuss that choice and reach a consensus. Meanwhile, I reverted your changes in a1b6c5f5d6f22088dc017552c94d133128059ad8.

Comment (rht) [inline pkgs/development/compilers/ghc/head.nix]: noted: I will push the case specific to GHC and GHCjs in a new PR
